### PR TITLE
add the model based policy gradient method

### PR DESCRIPTION
**Pull request by JiahaoYao**

### CODE REVIEW

**Comment by JiahaoYao** [comment]
how to overcome this?

File: recirq/optimize/mpg.py [hunk 1]
```diff
@@ -0,0 +1,390 @@
+# Copyright 2018 The Cirq Developers
```
Comment (mpharrigan) [inline recirq/optimize/mpg.py:1]: nit: 2020

File: recirq/optimize/mpg_test.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+
+    assert isinstance(result.x, np.ndarray)
+    assert isinstance(result.fun, float)
+    assert result.nit == 15
+
+
+def test_model_policy_gradient_limited_evaluations():
+    x0 = np.random.randn(10)
+    sample_radius = 1e-1
+    rate = 1e-1
```
Comment (mpharrigan) [inline recirq/optimize/mpg_test.py:85]: sample_radius and rate are unused. were they supposed to be used for something?

File: recirq/optimize/mpg_test.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+
+def sum_of_squares(x):
+    return np.sum(x**2).item()
+
+
+def test_model_policy_gradient():
+    x0 = np.random.randn(5)
+    result = model_policy_gradient(sum_of_squares,
+                                    x0,
+                                    learning_rate= 1e-1,
```
Comment (mpharrigan) [inline recirq/optimize/mpg_test.py:28]: formatting: omit the space after `=`

File: recirq/optimize/mpg_test.py [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+
+def sum_of_squares(x):
+    return np.sum(x**2).item()
+
+
+def test_model_policy_gradient():
+    x0 = np.random.randn(5)
+    result = model_policy_gradient(sum_of_squares,
+                                    x0,
```
Comment (mpharrigan) [inline recirq/optimize/mpg_test.py:27]: formatting: align these indents

File: recirq/optimize/mpg_test.py [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+import numpy as np
+
+from recirq.optimize.mpg import model_policy_gradient
+
+
+def sum_of_squares(x):
+    return np.sum(x**2).item()
+
+
+def test_model_policy_gradient():
```
Comment (mpharrigan) [inline recirq/optimize/mpg_test.py:24]: these tests take ~45 seconds on my machine, which is too long for a unit test. Can we provide a more concise check of the functionality?

File: recirq/optimize/mpg.py [hunk 6]
```diff
@@ -0,0 +1,390 @@
...
+        [("poly", PolynomialFeatures(degree=2)), ("linear_model", linear_model),]
+    )
+    shifted_xs = [(x - xopt) for x in xs]
+    model = model.fit(shifted_xs, ys)
+    return model
+
+
+def clip_by_norm(t, clip_norm):
+    """Clips tensor values to a maximum L2-norm.
+    returns a tensor of the same type and shape as `t` with its values set to:
```
Comment (mpharrigan) [inline recirq/optimize/mpg.py:56]: nit: newline between the summary line of the docstring and additional description

File: recirq/optimize/mpg.py [hunk 7]
```diff
@@ -0,0 +1,390 @@
...
+        Returns: 
+            a class of the schedule
+        """
+        self._v = value
+
+    def value(self, t):
+        """Return the value of the schedule at time step t 
+
+        Args: 
+            t: the time step for the schedule
```
Comment (mpharrigan) [inline recirq/optimize/mpg.py:88]: note that `t` is ignored here

File: recirq/optimize/mpg.py [hunk 8]
```diff
@@ -0,0 +1,390 @@
...
+        """Return the value of the schedule at time step t 
+
+        Args: 
+            t: the time step for the schedule 
+
+        Returns: 
+            the schedule value
+        """
+        m = t / self.decay_steps
+        if self.staircase == True:
```
Comment (mpharrigan) [inline recirq/optimize/mpg.py:131]: remove `== True`, this is a boolean

File: recirq/optimize/mpg.py [hunk 9]
```diff
@@ -0,0 +1,390 @@
...
+        Args: 
+            t: the time step for the schedule 
+
+        Returns: 
+            the schedule value
+        """
+        return self._v
+
+
+class ExponentialSchedule(object):
```
Comment (mpharrigan) [inline recirq/optimize/mpg.py:96]: nit: don't need `(object)`

File: recirq/optimize/mpg.py [hunk 10]
```diff
@@ -0,0 +1,390 @@
...
+        clip_norm: The scalar to give a maximal upper bound for the L2-norm of the tensor.
+
+    Returns:
+        the tensor with the same shape of the input, but with norm clipped.
+    """
+    l2_norm = np.linalg.norm(t, 2)
+    return t * clip_norm / np.maximum(l2_norm, clip_norm)
+
+
+class ConstantSchedule(object):
```
Comment (mpharrigan) [inline recirq/optimize/mpg.py:70]: nit: don't need `object`.

Consider making this a `dataclass` with `frozen=True`, especially if you're going to use it as a default argument for a function (default arguments should not be mutable)

File: recirq/optimize/mpg.py (hunk 9 above)
Comment (mpharrigan) [reply recirq/optimize/mpg.py:96]: this can probably be a dataclass as well

File: recirq/optimize/mpg.py [hunk 11]
```diff
@@ -0,0 +1,390 @@
...
+
+        Args:
+            grad: the gradient computed at the current update.  
+            x: the current value of the parameters. 
+
+        Returns:
+            the updated parameter values. 
+        """
+        if self.step == 0:
+            self.m = np.zeros(len(x))
```
Comment (mpharrigan) [inline recirq/optimize/mpg.py:173]: you shouldn't initialize class members outside of `__init__`

File: recirq/optimize/mpg.py [hunk 12]
```diff
@@ -0,0 +1,390 @@
...
+                # use the model for prediction
+                new_ys = model.predict(new_xs - current_x)
+                reward = [-y for y in new_ys]
+            except:
+                pass
+
+        reward = np.array(reward)
+
+        # stats
+        reward_mean = np.mean(reward)
```
Comment (mpharrigan) [inline recirq/optimize/mpg.py:349]: unused

File: recirq/optimize/mpg.py [hunk 13]
```diff
@@ -0,0 +1,390 @@
...
+                model = _get_quadratic_model(model_xs, model_ys, x)
+
+                # get samples (from model)
+                z = np.random.randn(batch_size_model, n)
+                new_xs = sigma * z + current_x
+
+                # use the model for prediction
+                new_ys = model.predict(new_xs - current_x)
+                reward = [-y for y in new_ys]
+            except:
```
Comment (mpharrigan) [inline recirq/optimize/mpg.py:343]: this is too broad of an exception catch. Please only catch errors you anticipate so this doesn't silently swallow unrelated, "real" errors

File: recirq/optimize/mpg.py [hunk 14]
```diff
@@ -0,0 +1,390 @@
...
+            # safer way without the `SVD` not converging
+            try:
+                for x, y in zip(known_xs, known_ys):
+                    if np.linalg.norm(x - current_x) < radius_coeff * max_radius:
+                        model_xs.append(x)
+                        model_ys.append(y)
+                model = _get_quadratic_model(model_xs, model_ys, x)
+
+                # get samples (from model)
+                z = np.random.randn(batch_size_model, n)
```
Comment (mpharrigan) [inline recirq/optimize/mpg.py:337]: It's usually useful to have deterministic versions of stochastic functions. Consider accepting a numpy random state as an optional argument. see `cirq.parse_random_state` for a helpful utility function and https://cirq.readthedocs.io/en/stable/generated/cirq.RANDOM_STATE_OR_SEED_LIKE.html#cirq.RANDOM_STATE_OR_SEED_LIKE for more details

File: recirq/optimize/mpg.py [hunk 15]
```diff
@@ -0,0 +1,390 @@
...
+            the schedule value
+        """
+        m = t / self.decay_steps
+        if self.staircase == True:
+            m = np.floor(m)
+
+        return self.learning_rate * self.decay_rate ** m
+
+
+class Adam:
```
Comment (mpharrigan) [inline recirq/optimize/mpg.py:137]: why is this a class? just make it a function

**Comment by JiahaoYao** [comment]
Hi @mpharrigan, received and thanks for your careful comments! Will fix them as soon as possible!

File: recirq/optimize/mpg.py (hunk 9 above)
Comment (JiahaoYao) [reply recirq/optimize/mpg.py:96]: Good point! Thanks!

File: recirq/optimize/mpg.py (hunk 11 above)
Comment (JiahaoYao) [reply recirq/optimize/mpg.py:173]: agreed!

File: recirq/optimize/mpg.py (hunk 13 above)
Comment (JiahaoYao) [reply recirq/optimize/mpg.py]: make it narrow

File: recirq/optimize/mpg_test.py (hunk 5 above)
Comment (JiahaoYao) [reply recirq/optimize/mpg_test.py:23]: lower the criterion

File: recirq/optimize/mpg.py (hunk 15 above)
Comment (mpharrigan) [reply recirq/optimize/mpg.py]: I meant `Adam`. `ExponentialSchedule` makes sense as a class since it closes over the hyperparameters (learning rate et. al.) and follows the (implicit) Schedule interface

File: recirq/optimize/mpg.py [hunk 16]
```diff
@@ -67,56 +74,31 @@ def clip_by_norm(t, clip_norm):
     return t * clip_norm / np.maximum(l2_norm, clip_norm)
 
 
-class ConstantSchedule(object):
```
Comment (mpharrigan) [inline recirq/optimize/mpg.py]: Why did you remove `ConstantSchedule`?

File: recirq/optimize/mpg.py [hunk 17]
```diff
@@ -128,64 +110,60 @@ def value(self, t):
...
         return self.learning_rate * self.decay_rate ** m
 
 
-class Adam:
-    def __init__(
-        self, lr_schedule=ConstantSchedule(0.001), b1=0.9, b2=0.999, eps=10 ** -8
-    ):
-        """The optimizer Adam: a method for stochastic gradient descent  
+def Adam_update(
+    grad,
```
Comment (mpharrigan) [inline recirq/optimize/mpg.py:120]: type annotations would help here

File: recirq/optimize/mpg.py [hunk 18]
```diff
@@ -128,64 +110,60 @@ def value(self, t):
...
 
         return self.learning_rate * self.decay_rate ** m
 
 
-class Adam:
-    def __init__(
-        self, lr_schedule=ConstantSchedule(0.001), b1=0.9, b2=0.999, eps=10 ** -8
-    ):
-        """The optimizer Adam: a method for stochastic gradient descent  
+def Adam_update(
```
Comment (mpharrigan) [inline recirq/optimize/mpg.py:119]: I know the original authors style it as "Adam", but I think we should follow PEP8 here and call the function `adam_update`. 

An additional question: do we expect users to use this function or will it always be called from `model_policy_gradient`? That is -- can we make this private (by naming it `_adam_update`)?

File: recirq/optimize/mpg.py (hunk 16 above)
Comment (JiahaoYao) [reply recirq/optimize/mpg.py]: oh, because it is not used in the code? Shall I put it back?

File: recirq/optimize/mpg.py (hunk 17 above)
Comment (JiahaoYao) [reply recirq/optimize/mpg.py:120]: good point!

File: recirq/optimize/mpg.py (hunk 18 above)
Comment (JiahaoYao) [reply recirq/optimize/mpg.py:119]: yes, that is right. I have fixed that.

File: recirq/optimize/mpg.py (hunk 15 above)
Comment (JiahaoYao) [reply recirq/optimize/mpg.py]: agreed!

File: recirq/optimize/mpg.py (hunk 7 above)
Comment (JiahaoYao) [reply recirq/optimize/mpg.py]: so, the constant schedule is supposed to ignore the argument `t` because it maps everything to the constant. Can I use `t` as the argument?

File: recirq/optimize/mpg.py (hunk 16 above)
Comment (mpharrigan) [reply recirq/optimize/mpg.py]: Well it was there before so I was wondering why you removed it :) It's up to you and what use cases you want to support

Comment (JiahaoYao) [reply recirq/optimize/mpg.py]: ha, I find the schedule is not exposed to the "outside". so, previous constant schedule can not be used. 
But, in practice, it always good to use the decay schedule. If anyone finds the constant schedule is better, I can then put it back.

**Comment by JiahaoYao** [comment]
> Looks great! thanks again. If you do want to put constantschedule back in, feel free to do so in a follow-up pr

Yes, thanks for approving that!
